packages/editor/src/lib/editor/managers/SnapManager.ts

```diff
@@ -440,180 +457,92 @@ export class SnapManager {
-
-		const pageHeadSegments = spline.segments
-			.slice(0, Math.max(0, segmentNumber - ignoreCount))
-			.map((s) => Matrix2d.applyToPoints(pageTransform, s.lut))
-
-		const pageTailSegments = spline.segments
-			.slice(segmentNumber + ignoreCount)
-			.map((s) => Matrix2d.applyToPoints(pageTransform, s.lut))
-
-		return this.snapHandleTranslate({
-			handlePoint: handlePoint,
-			additionalOutlines: [...pageHeadSegments, ...pageTailSegments],
+	@computed get outlinesInPageSpace() {
+		return this.snappableShapes.map(({ id, isClosed }) => {
+			const outline = deepCopy(this.editor.getOutlineById(id))
+			if (isClosed) outline.push(outline[0])
+			const pageTransform = this.editor.getPageTransformById(id)
+			if (!pageTransform) throw Error('No page transform')
+			return Matrix2d.applyToPoints(pageTransform, outline)
 		})
 	}
 
-	// for a handle:
-	// - find the nearest snap point from all non-selected shapes
-	// - return the nudge vector to snap to that point
-	// note: this happens within page space
-	snapHandleTranslate({
+	getSnappingHandleDelta({
 		handlePoint,
-		additionalOutlines = [],
+		additionalSegments,
 	}: {
 		handlePoint: Vec2d
-		additionalOutlines?: Vec2dModel[][]
-	}): SnapData {
-		// Get the (page-space) outlines of the shapes that are not in the selection
-		const visibleShapesNotInSelection = this.visibleShapesNotInSelection
-		const pageOutlines = []
-		for (const visibleShape of visibleShapesNotInSelection) {
-			const shape = this.editor.getShapeById(visibleShape.id)!
-
-			if (shape.type === 'text' || shape.type === 'icon') {
-				continue
-			}
-
-			const outline = deepCopy(this.editor.getOutlineById(visibleShape.id))
-
-			const isClosed = this.editor.getShapeUtil(shape).isClosed?.(shape)
-
-			if (isClosed) {
-				outline.push(outline[0])
-			}
-
-			pageOutlines.push(
-				Matrix2d.applyToPoints(this.editor.getPageTransformById(shape.id)!, outline)
-			)
-		}
+		additionalSegments: Vec2d[][]
+	}): Vec2d | null {
+		const { outlinesInPageSpace, snapThreshold } = this
 
 		// Find the nearest point that is within the snap threshold
-		let minDistance = this.snapThreshold
+		let minDistance = snapThreshold
 		let nearestPoint: Vec2d | null = null
-		for (const outline of [...pageOutlines, ...additionalOutlines]) {
+		let C: Vec2dModel, D: Vec2dModel, nearest: Vec2d, distance: number
+
+		const allSegments = [...outlinesInPageSpace, ...additionalSegments]
+		for (const outline of allSegments) {
 			for (let i = 0; i < outline.length - 1; i++) {
-				const C = outline[i]
-				const D = outline[i + 1]
+				C = outline[i]
+				D = outline[i + 1]
+
+				nearest = Vec2d.NearestPointOnLineSegment(C, D, handlePoint)
+				distance = Vec2d.Dist(handlePoint, nearest)
 
-				const distance = Vec2d.DistanceToLineSegment(C, D, handlePoint)
 				if (isNaN(distance)) continue
 				if (distance < minDistance) {
 					minDistance = distance
-					nearestPoint = Vec2d.NearestPointOnLineSegment(C, D, handlePoint)
```

Comment:
DistanceToLineSegment also calls NearestPointOnLineSegment, so we can just find that point and then measure the distance ourselves.